File: backend/app/services/completeness.py
"""充実度スコア（いいね経済の中核）。
SSoT は本ファイルと frontend/src/lib/completeness.ts の二重実装。
数式を変える時は必ず両方を同時に変更すること。
スコア = 写真(最大15) + 自己紹介(最大25) + 雑多(最大60) = 最大100。"""

MISC_FIELDS: tuple[str, ...] = (
    "height_cm", "body_type", "blood_type", "sibling_rank", "languages",
    "campus", "housing", "commute_time", "commute_means", "second_lang",
    "marriage_intent", "preferred_age_band", "drinking", "smoking", "mbti",
    "hometown", "free_slots",
)
_ARRAY_MISC: frozenset[str] = frozenset({"languages", "commute_means"})

BIO_FULL_LEN = 350  # この文字数で bio 満点(25)
PHOTO_CAP = 4       # 4枚で写真満点(15)・以降加算なし


def _is_filled(field: str, value) -> bool:
    if value is None:
        return False
    if field in _ARRAY_MISC:
        return isinstance(value, list) and len(value) > 0
    if isinstance(value, str):
        return value.strip() != ""
    return True  # height_cm(smallint) 等


def _photo_points(photo_count: int) -> float:
    # 1枚=0 / 2枚=5 / 3枚=10 / 4枚以上=15
    return max(0, min(photo_count, PHOTO_CAP) - 1) * 5.0


def _bio_points(bio_len: int) -> float:
    # 25 * (len/350)^1.5（350字で頭打ち25）
    ratio = min(bio_len, BIO_FULL_LEN) / BIO_FULL_LEN
    return 25.0 * (ratio ** 1.5)


def compute_completeness(profile: dict, photo_count: int) -> dict:
    """profile: profiles 行の dict。photo_count: 本人の profile_images 件数（rejected 除く）。
    戻り値: score(0-100 float) と内訳・未入力雑多リスト（伸び幅アドバイス用）。"""
    bio_len = len((profile.get("bio") or "").strip())
    filled = sum(1 for f in MISC_FIELDS if _is_filled(f, profile.get(f)))
    unfilled = [f for f in MISC_FIELDS if not _is_filled(f, profile.get(f))]

    photo_pts = _photo_points(photo_count)
    bio_pts = _bio_points(bio_len)
    misc_pts = 60.0 * filled / len(MISC_FIELDS)
    score = photo_pts + bio_pts + misc_pts  # 0-100

    return {
        "score": round(score, 1),
        "photo_count": photo_count,
        "photo_points": round(photo_pts, 1),
        "bio_length": bio_len,
        "bio_points": round(bio_pts, 1),
        "misc_filled": filled,
        "misc_total": len(MISC_FIELDS),
        "misc_points": round(misc_pts, 1),
        "unfilled_misc": unfilled,
    }

File: frontend/src/lib/completeness.ts
// 充実度スコア（いいね経済の中核）。
// SSoT は本ファイルと backend/app/services/completeness.py の二重実装。
// 数式を変える時は必ず両方を同時に変更すること。

export const MISC_FIELDS = [
  'height_cm', 'body_type', 'blood_type', 'sibling_rank', 'languages',
  'campus', 'housing', 'commute_time', 'commute_means', 'second_lang',
  'marriage_intent', 'preferred_age_band', 'drinking', 'smoking', 'mbti',
  'hometown', 'free_slots',
] as const;

const ARRAY_MISC = new Set<string>(['languages', 'commute_means']);
export const BIO_FULL_LEN = 350;
export const PHOTO_CAP = 4;

function isFilled(field: string, value: unknown): boolean {
  if (value === null || value === undefined) return false;
  if (ARRAY_MISC.has(field)) return Array.isArray(value) && value.length > 0;
  // 全0文字列（空コマ）は未入力扱い（"1"を1つ以上含む場合のみ埋まりとみなす）
  if (field === 'free_slots') return typeof value === 'string' && value.includes('1');
  if (typeof value === 'string') return value.trim() !== '';
  return true;
}

export function photoPoints(photoCount: number): number {
  return Math.max(0, Math.min(photoCount, PHOTO_CAP) - 1) * 5;
}

export function bioPoints(bioLen: number): number {
  const ratio = Math.min(bioLen, BIO_FULL_LEN) / BIO_FULL_LEN;
  return 25 * Math.pow(ratio, 1.5);
}

export interface Completeness {
  score: number;        // 0-100
  photoCount: number;
  photoPoints: number;
  bioLength: number;
  bioPoints: number;
  miscFilled: number;
  miscTotal: number;
  miscPoints: number;
  unfilledMisc: string[];
}

export const SAME_SEX_UNLOCK = 70
export const MALE_BONUS_THRESHOLD = 80

export function sendRegime(
  gender: string | null | undefined,
  interestIn: string | null | undefined,
): string {
  if (gender === 'female' && interestIn === 'male') return 'female_unlimited'
  if (gender === 'male' && interestIn === 'female') return 'male_hetero'
  if (gender != null && gender === interestIn) return 'same_sex'
  return 'female_unlimited'
}

export function computeCompleteness(
  profile: Record<string, unknown>,
  photoCount: number,
): Completeness {
  const bioLen = typeof profile.bio === 'string' ? profile.bio.trim().length : 0;
  const filled = MISC_FIELDS.filter((f) => isFilled(f, profile[f])).length;
  const unfilled = MISC_FIELDS.filter((f) => !isFilled(f, profile[f]));
  const pPts = photoPoints(photoCount);
  const bPts = bioPoints(bioLen);
  const mPts = (60 * filled) / MISC_FIELDS.length;
  return {
    score: Math.round((pPts + bPts + mPts) * 10) / 10,
    photoCount,
    photoPoints: Math.round(pPts * 10) / 10,
    bioLength: bioLen,
    bioPoints: Math.round(bPts * 10) / 10,
    miscFilled: filled,
    miscTotal: MISC_FIELDS.length,
    miscPoints: Math.round(mPts * 10) / 10,
    unfilledMisc: [...unfilled],
  };
}
